### ISSUE TITLE
How small object optimizations work

### ISSUE THREAD

**Issue by lefticus** [lefticus/cpp_weekly#98]
`std::any`, `std::function` and `std::string` each have optimizations that avoid heap allocations for the case of "small" objects. How do these optimizations work?

**Comment by lefticus** [comment]
Here's a demo: https://compiler-explorer.com/z/a3q6x3Eas

**Comment by lefticus** [comment]
Live version: https://compiler-explorer.com/z/PqKhMvqWE

**Comment by lefticus** [comment]
Coming in Ep430

**Comment by joaotavora** [comment]
This is great, but clang compiles about a lot of stuff.  Is it channel policy to not care about portability in general?

**Comment by lefticus** [comment]
> This is great, but clang compiles about a lot of stuff. Is it channel policy to not care about portability in general?

This is a bizarrely antagonistic way of asking a question about portability. If you have watched many of my other videos or my conference talks, or look into my starter project, or at any of my open source projects, you should know that I take portability seriously.

I simply don't take the time to try every code sample with every compiler, and sometimes things fall through the cracks.

Here is a version that does work on clang, gcc and MSVC: https://compiler-explorer.com/z/hY7Goxa93

**Comment by joaotavora** [comment]
> This is a bizarrely antagonistic way of asking a question about portability. 

ahahaha, bizarre antagonist is 100% my next alter-ego.  

>  If you have watched many of my other videos or my conference talks, or look into my starter project, or at any of my open source projects, you should know that I take portability seriously.

Of course I do.  I did all those 3 things.  I watch lots of your stuff.  Your work is awesome and much admired. I just figured maybe the point of this video or this channel in particular or this example in particular was not to care much about portability, but just illustrate some SSO ideas.  I found this video so interesting I started following along.  My gcc compiled it perfectly, but clangd constantly complained about stuff, so I asked.  Sorry I came across antagonistic, didn't introduce myself, etc.

> Here is a version that does work on clang, gcc and MSVC: https://compiler-explorer.com/z/hY7Goxa93

superb, many thanks. Also, is GitHub the right place for questions about your videos?
